File: backend/src/main/java/de/codeministry/leadgen/offer/ShortlistEntry.java
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Copyright 2026 Marcello Muscara (codeministry)
 *
 * Licensed under the Apache License, Version 2.0. You may obtain a copy of the
 * License at http://www.apache.org/licenses/LICENSE-2.0
 */
package de.codeministry.leadgen.offer;

import de.codeministry.leadgen.content.ContentBlock;

import java.util.List;

/**
 * One *project* on the shortlist, which is not the same thing as one listing.
 *
 * <p>{@code sources} holds every portal advertising it, collapsed by deduplication — 12.3
 * % of the measured corpus reaches the pipeline more than once, and a shortlist that shows
 * the same project three times is a shortlist nobody finishes reading.
 *
 * <p>{@code content} is the advert read into blocks, and it is <b>populated by the detail
 * query alone</b>. The list uses the same row mapper, so a block list carried along with it
 * would put a second copy of every advert into a response the repository already measures in
 * megabytes — for a column no card renders. Empty therefore means two things that want the
 * same answer: this is the list, or nothing has read the advert that way yet. Both fall back
 * to {@code offer.fullText}.
 */
public record ShortlistEntry(
    OfferView offer,
    OfferScoreView score,
    OfferFlags flags,
    List<OfferSourceRef> sources,
    List<ContentBlock> content) {
}

File: frontend/src/app/core/model/shortlist-entry.ts
import {ContentBlock, Offer, OfferFlags, OfferSource} from './offer';
import {Score} from './score';

/**
 * One *project* on the shortlist, which is not the same thing as one listing:
 * `sources` holds every portal advertising it, collapsed by deduplication. The
 * concept has no name for this yet, so it gets one here.
 *
 * `content` is the advert read into blocks, and only the detail endpoint fills it — the list
 * shares the server's row mapper and would otherwise carry a second copy of every advert for
 * something no card renders. So empty means two things that want the same answer: this came
 * from the list, or nothing has read the advert that way yet. Both fall back to
 * `offer.fullText`.
 */
export interface ShortlistEntry {
    readonly offer: Offer;
    readonly score: Score;
    readonly flags: OfferFlags;
    readonly sources: readonly OfferSource[];
  readonly content: readonly ContentBlock[];
}
